File: java/cours/src/session2/td/heritageMultiple/SuccRecursif.java
package session2.td.heritageMultiple;

import session1.td.Nat;

public class SuccRecursif extends EtatSucc implements AlgebreNatRecursifSuccesseur{

	public SuccRecursif(Nat rep) {
		super(rep);
	}
	
	@Override
	public Nat creerZero() {
		return new ZeroRecursif();
	}
	
	@Override
	public Nat creerSuccesseur(Nat predecesseur) {
		return new SuccRecursif(predecesseur);
	}
	
	@Override
	public String toString() {
		return Integer.toString(this.val());
	}

	@Override
	public boolean equals(Object obj) {
		if (!(obj instanceof Nat))
			return false;
		Nat x = (Nat) obj;
		return this.val() == x.val();
	}
}

File: typescript/session2/heritageMultiple/SuccRecursif.ts
import {EtatSucc} from "./EtatSucc";
import {AlgebreNatRecursifSuccesseur} from "./AlgebreNatRecursifSuccesseur";
import {Nat} from "../../session1/td/Nat/Nat";
import {ZeroRecursif} from "./ZeroRecursif";

export class SuccRecursif extends EtatSucc, AlgebreNatRecursifSuccesseur{

    constructor(rep : Nat) {
        super(rep);
    }

    creerZero() : Nat{
        return new ZeroRecursif();
    }

    creerSuccesseur(predecesseur : Nat) : Nat {
        return new SuccRecursif(predecesseur);
    }

    toString() : string{
        return '' + this.val();
    }

    equals(obj:Object) : boolean{
        if (!(obj instanceof Nat))
            return false;
        let x = <Nat> obj;
        return this.val() == x.getVal();
    }
}
